core/src/main/java/org/apache/iceberg/IncrementalDataTableScan.java

```diff
@@ -19,55 +19,43 @@
 
 package org.apache.iceberg;
 
-import java.util.Collection;
 import java.util.List;
 import java.util.Set;
-import org.apache.iceberg.expressions.Expression;
 import org.apache.iceberg.io.CloseableIterable;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
 import org.apache.iceberg.relocated.com.google.common.collect.FluentIterable;
-import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
 import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
 import org.apache.iceberg.relocated.com.google.common.collect.Lists;
 import org.apache.iceberg.relocated.com.google.common.collect.Sets;
 import org.apache.iceberg.util.SnapshotUtil;
 import org.apache.iceberg.util.ThreadPools;
 
 class IncrementalDataTableScan extends DataTableScan {
-  private long fromSnapshotId;
-  private long toSnapshotId;
-
-  IncrementalDataTableScan(TableOperations ops, Table table, Schema schema, Expression rowFilter,
-                           boolean ignoreResiduals, boolean caseSensitive, boolean colStats,
-                           Collection<String> selectedColumns, ImmutableMap<String, String> options,
-                           long fromSnapshotId, long toSnapshotId) {
-    super(ops, table, null, schema, rowFilter, ignoreResiduals, caseSensitive, colStats, selectedColumns, options);
-    validateSnapshotIds(table, fromSnapshotId, toSnapshotId);
-    this.fromSnapshotId = fromSnapshotId;
-    this.toSnapshotId = toSnapshotId;
+
+  IncrementalDataTableScan(TableOperations ops, Table table, Schema schema, TableScanContext context) {
+    super(ops, table, schema, context.useSnapshotId(null));
```

Comment:
Doesn't the snapshot ID start off as null? It seems like we don't need to set it here.

Reply:
This change was done following the previous code that did pass `null` in the `super` constructor. `TableScanContext` does start as null, but there's no guarantee that it's still `null` in this part since it's only receiving a context as argument, not creating one from scratch, right?

Reply:
The guarantee that the snapshot ID is currently null is the check done in the refinement methods that create an incremental scan. I think those validations should be done in the context because the context should be responsible for knowing what is a valid scan. If a table doesn't support incremental scanning, then it can have additional validations, but anything common should be in the context.

That would allow you to not set the snapshot to null here. We may also want to get rid of `IncrementalDataTableScan`, though probably not in this PR. I think the purpose of it was to minimize changes to `BaseTableScan`. But this refactor makes that much less necessary.